src/service/xhr-impl.js

```diff
@@ -344,17 +344,26 @@ function isRetriable(status) {
 /**
  * Returns the response if successful or otherwise throws an error.
  * @paran {!FetchResponse} response
- * @return {!FetchResponse}
+ * @return {!Promise<!FetchResponse>}
  */
 function assertSuccess(response) {
-  if (!(response.status >= 200 && response.status < 300)) {
-    const err = user.createError(`HTTP error ${response.status}`);
-    if (isRetriable(response.status)) {
-      err.retriable = true;
+  return new Promise((resolve, reject) => {
+    if (!(response.status >= 200 && response.status < 300)) {
+      const err = user.createError(`HTTP error ${response.status}`);
+      if (isRetriable(response.status)) {
+        err.retriable = true;
+      }
+      return response.json().then(json => {
```

Comment:
(You'll probably move this to a separate PR, but...) Before attempting to do `json()` here, we need to check that the `Content-Type` is actually JSON. The error response doesn't have to follow the same protocol as the successful response and typically doesn't do so.